src/fairseq2/datasets/nllb.py

```diff
@@ -0,0 +1,409 @@
+        """
+        self.dataset_name = dataset_name
+        self.data_dir = data_dir
+        self.splits = splits
+
+    @finaloverride
+    def read(
+        self,
+        split: str,
+        tokenizer: TextTokenizer,
+        gang: Gang,
+        *,
+        max_seq_len: int = 512,
+        max_num_tokens: int = 2048,
+        shuffle_window_size: int = 10_000,
+        eval_batch_size: int = 32,
+        num_prefetch: int = 10,
+        lang_pairs: Optional[Sequence[LangPair]] = None,
+    ) -> DataPipeline[Seq2SeqBatch]:
+        split_names = []
+
+        for split_name in self.splits.keys():
+            if split_name == split or split_name.startswith(split + "_"):
+                split_names.append(split_name)
+
+        if not split_names:
+            raise ValueError(
+                f"`split` must be a valid split name, but the {self.dataset_name} dataset has no split named '{split}'."
+            )
+
+        splits = {}
+
+        # Extract the language pairs along with their sizes from the requested
+        # split or splits.
+        for split_name in split_names:
+            split_lang_pairs = self.splits[split_name]
+
+            if lang_pairs is None:
+                lang_pairs_to_read = split_lang_pairs
+            else:
+                lang_pairs_to_read = {}
+
+                for lang_pair in lang_pairs:
+                    try:
+                        lang_pairs_to_read[lang_pair] = split_lang_pairs[lang_pair]
+                    except KeyError:
+                        raise ValueError(
+                            f"All language pairs in `lang_pairs` must be available in the dataset, but '{lang_pair}' is not found in the '{split}' split of the {self.dataset_name} dataset."
+                        )
+
+            splits[split_name] = lang_pairs_to_read
+
+        train = split == "train" or split.startswith("train_")
+
+        factory = _NllbDataPipelineFactory(
+            self.dataset_name,
+            self.data_dir,
+            tokenizer,
+            splits,
+            gang,
+            train,
+            max_seq_len,
+            max_num_tokens,
+            shuffle_window_size,
+            eval_batch_size,
+            num_prefetch,
+        )
+
+        return factory()
+
+    @finaloverride
+    def get_splits(self) -> List[str]:
+        return list(self.splits.keys())
+
+    @finaloverride
+    def get_lang_pairs(self, split: str) -> List[LangPair]:
+        try:
+            return list(self.splits[split].keys())
+        except KeyError:
+            raise ValueError(
+                f"`split` must be a valid split name, but the {self.dataset_name} dataset has no split named '{split}'."
+            )
+
+
+class _LangPairExample(TypedDict):
+    """Represents a line read from a language pair file."""
+
+    split: CString
+    lang_pair: CString
+    line_number: int
+    source_tokens: Tensor
+    target_tokens: Tensor
+
+
+class _NllbDataPipelineFactory:
```

Comment:
Does the `_` in the name added to suggest these are internal classes ?

Reply:
Yes, any module-scope type, function, or variable that is prefixed with an underscore indicates "internal". For class method and variables depending on the use case it might mean either "private" or "protected". If it is "protected", I typically leave a comment like `# protected`.